runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java

```diff
@@ -286,23 +288,32 @@ protected boolean followBackpointers(final UUID streamId,
             // Now calculate the next address
             // Try using backpointers first
 
-            log.trace("followBackPointers: calculate the next address");
+            log.trace("followBackpointers: calculate the next address");
 
             if (!runtime.getParameters().isBackpointersDisabled() && d.hasBackpointer(streamId)) {
                 long tmp = d.getBackpointer(streamId);
-                log.trace("followBackPointers: backpointer points to {}", tmp);
+                log.trace("followBackpointers: backpointer points to {}", tmp);
                 // if backpointer is a valid log address or Address.NON_EXIST
                 // (beginning of the stream), do not single step back on the log
                 if (Address.isAddress(tmp) || tmp == Address.NON_EXIST) {
                     currentAddress = tmp;
                     singleStep = false;
+                    if (!startingSingleStep) {
+                        // A started single step period finishes here, refresh flag for next cycle.
+                        log.info("followBackpointers[{}]: Found backpointer for this stream at address {}. Stop single step downgrade.");
+                        startingSingleStep = true;
+                    }
                 }
             }
 
             if (singleStep) {
+                if (startingSingleStep) {
+                    startingSingleStep = false;
+                    log.info("followBackpointers[{}]: Found hole at address {}. Starting single step downgrade.", this, currentAddress);
+                }
                 // backpointers failed, so we're
                 // downgrading to a linear scan
-                log.warn("followBackPointers[{}]: downgrading to single step, found hole at {}", this, currentAddress);
+                log.trace("followBackpointers[{}]: downgrading to single step, found hole at {}", this, currentAddress);
```

Comment:
Do we need `Trace`? Or you think `Debug` would be fine.

Reply:
@zalokhan I was actually not sure on this one. I was trying to be consistent with the other logging level/messages we have in this same method. For instance, printing the backpointer for each entry is on trace level, which is kind of the same we're doing here but for entries not belonging to this stream. But if there is a strong reason to move it to debug, that is fine also, can you think of any? The only thing I can think of, is the fact that we'd typically want to debug this and not change to trace level which brings many more data. Let me know what you think is better in this case.

Reply:
I agree. Let's keep it in Trace then.